src/ConnectedMode/QualityProfiles/OutOfDateQPFinder.cs

```diff
@@ -0,0 +1,111 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.Binding;
+using SonarLint.VisualStudio.Integration;
+using SonarQube.Client;
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.ConnectedMode.QualityProfiles
+{
+    internal interface IOutOfDateQPFinder
+    {
+        /// <summary>
+        /// When in Connected Mode, returns the languages for which the Quality Profile
+        /// needs to be updated
+        /// </summary>
+        Task<IEnumerable<Language>> GetAsync(CancellationToken cancellationToken);
+    }
+
+    [Export(typeof(IOutOfDateQPFinder))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class OutOfDateQPFinder : IOutOfDateQPFinder
+    {
+        private readonly IConfigurationProvider configProvider;
+        private readonly ISonarQubeService sonarQubeService;
+        private readonly ILogger logger;
+
+        [ImportingConstructor]
+        public OutOfDateQPFinder(IConfigurationProvider configProvider,
+            ISonarQubeService sonarQubeService,
+            ILogger logger)
+        {
+            this.configProvider = configProvider;
+            this.sonarQubeService = sonarQubeService;
+            this.logger = logger;
+        }
+
+        public async Task<IEnumerable<Language>> GetAsync(CancellationToken cancellationToken)
+        {
+            var languagesToUpdate = new List<Language>();
+
+            var config = configProvider.GetConfiguration();
+            foreach (var profile in config.Project.Profiles)
+            {
+                logger.LogQPVerbose(profile.Key, "Checking if Quality Profile has changed");
+                if (await IsUpdateRequiredAsync(config, profile, cancellationToken))
+                {
+                    languagesToUpdate.Add(profile.Key);
+                    logger.LogQPVerbose(profile.Key, "Quality profile update is required");
+                }
+                else
+                {
+                    logger.LogQPVerbose(profile.Key, "Quality profile is up-to-date");
+                }
+            }
+
+            return languagesToUpdate;
+        }
+
+        private async Task<bool> IsUpdateRequiredAsync(BindingConfiguration config, KeyValuePair<Language, ApplicableQualityProfile> profile, CancellationToken cancellationToken)
+        {
+            Debug.Assert(profile.Value != null, "Not expecting profile to be null. Language: " + profile.Key);
+
+            var serverProfileInfo = await sonarQubeService.GetQualityProfileAsync(config.Project.ProjectKey,
```

Comment:
There's a search method that returns the list of all QP for the project. Using that would mean that if we find out we actually want to update a QP, we would still have to call Get for that specific QP, but considering the fact that QPs don't change often, on average we would make 1 request per timed update, rather than making 1 for each language

Reply:
Actually, since we're not using the info we fetched here and instead change the QP later (presumably making another Get for each changed QP?) using the search method makes even more sense

Reply:
https://next.sonarqube.com/sonarqube/web_api/api/qualityprofiles/search